File: leetcode/0169_majority_element/Leetcode169.java
/**
 * Tags: Arrays, Hash table, Sorting, Divide and conquer, Counting
 * Difficulty: Easy
 *
 * Time complexity: O(N)
 * Space complexity: O(1)
 */

public class Leetcode169 {
    public int majorityElement(int[] nums) {
        int majority = nums[0]; // Initial candidate for majority
        int count = 1; // Start the counter at 1 because we assume the first element as majority

        for (int i = 1; i < nums.length; i++) {
            if (nums[i] == majority) {
                count++; // If current element matches the candidate, increment the count
            } else {
                count--; // Otherwise, decrement the count
                if (count == 0) {
                    majority = nums[i]; // Update majority candidate
                    count = 1; // Reset the counter
                }
            }
        }

        return majority;
    }
}

File: leetcode/0169_majority_element/leetcode169.js
var majorityElement = function (nums) {
	let count = 0;
	let candidate;
	for (let i = 0; i < nums.length; i++) {
		if (count === 0) {
			candidate = nums[i];
			count = 1;
		} else if (candidate === nums[i]) {
			count++;
		} else {
			count--;
		}
	}

	return candidate;
};
